src/frontend/lexer.rs

```diff
@@ -0,0 +1,259 @@
+    fn starts_integer(c: &char) -> bool {
+        c.is_ascii_digit()
+    }
+    fn parse_integer(&mut self) -> TokResult {
+        Ok(Token::Integer(self.collect_while(char::is_ascii_digit)))
+    }
+
+    // String --------------------------------------------------------------- //
+    fn starts_string(c: &char) -> bool {
+        *c == '"'
+    }
+    fn parse_string(&mut self) -> TokResult {
+        self.consume();
+        let contents = self.collect_while(|&c| c != '"');
+
+        if self.peek().is_none() {
+            // Missing the closing " on a string. Reached EOF
+            Err(Error::UnclosedString)
+        } else {
+            self.consume();
+            Ok(Token::String(contents))
+        }
+    }
+
+    // ---------------------------------------------------------------------- //
+    // Helper methods                                                         //
+    // ---------------------------------------------------------------------- //
+
+    // Single element ------------------------------------------------------- //
+    fn peek(&mut self) -> Option<&char> {
+        self.source.peek()
+    }
+    fn consume(&mut self) {
+        self.source.next();
+    }
+    fn accept(&mut self, t: Token) -> TokResult {
+        self.consume();
+        Ok(t)
+    }
+
+    // Multi element -------------------------------------------------------- //
+    fn collect_while<F: FnMut(&char) -> bool>(&mut self, matcher: F) -> String {
+        self.source.peeking_take_while(matcher).collect()
+    }
+    fn skip_while<F: FnMut(&char) -> bool>(&mut self, matcher: F) {
+        self.source.peeking_take_while(matcher).for_each(drop)
+    }
+}
+
+// -------------------------------------------------------------------------- //
+// Trait implementations                                                      //
+// -------------------------------------------------------------------------- //
+
+// Iterator ----------------------------------------------------------------- //
+impl<I> Iterator for Lexer<I>
+where
+    I: Iterator<Item = char>,
+{
+    type Item = TokResult;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next_token()
+    }
+}
+
+// From for iterators ------------------------------------------------------- //
+impl<I> From<I> for Lexer<I>
+where
+    I: Iterator<Item = char>,
+{
+    fn from(source: I) -> Self {
+        Self::new(source.peekable())
+    }
+}
+
+// From for string slices --------------------------------------------------- //
+impl<'a> From<&'a str> for Lexer<std::str::Chars<'a>> {
+    fn from(source: &'a str) -> Self {
+        Self::new(source.chars().peekable())
+    }
+}
+
+// -------------------------------------------------------------------------- //
+// Tests                                                                      //
+// -------------------------------------------------------------------------- //
+
+#[cfg(test)]
+mod tests {
```

Comment:
To challenge yourself: Could you try to get a CI/CD job setup and try running all tests with it.

Reply:
Alright, I'll try some time in the near future :)